### PR TITLE
[libc][math][c23] Add f16sqrtf C23 math function

### PR DESCRIPTION
**Pull request by overmighty**
Part of #95250.

### CODE REVIEW

File: libc/utils/MPFRWrapper/MPFRUtils.h [hunk 1]
```diff
@@ -271,7 +290,8 @@ class MPFRMatcher : public testing::Matcher<OutputType> {
 // types.
 template <Operation op, typename InputType, typename OutputType>
 constexpr bool is_valid_operation() {
-  return (Operation::BeginUnaryOperationsSingleOutput < op &&
+  return (op == Operation::Sqrt) ||
+         (Operation::BeginUnaryOperationsSingleOutput < op &&
```
Comment (overmighty) [inline libc/utils/MPFRWrapper/MPFRUtils.h:294]: All of the changes for the exhaustive test need cleanup before they can be merged, and especially this one.

File: libc/src/__support/FPUtil/sqrt.h [hunk 2]
```diff
@@ -13,7 +13,9 @@
 #include "src/__support/macros/properties/cpu_features.h"
 
 #if defined(LIBC_TARGET_ARCH_IS_X86_64) && defined(LIBC_TARGET_CPU_HAS_SSE2)
-#include "x86_64/sqrt.h"
+// #include "x86_64/sqrt.h"
+// TODO
+#include "generic/sqrt.h"
```
Comment (overmighty) [inline libc/src/__support/FPUtil/sqrt.h:18]: TODO

File: libc/src/__support/FPUtil/generic/sqrt.h [hunk 3]
```diff
@@ -17,6 +17,7 @@
 #include "src/__support/FPUtil/rounding_mode.h"
 #include "src/__support/common.h"
 #include "src/__support/uint128.h"
+#include <fenv.h>
```
Comment (overmighty) [inline libc/src/__support/FPUtil/generic/sqrt.h:20]: Oops.

File: libc/utils/MPFRWrapper/MPFRUtils.cpp [hunk 4]
```diff
@@ -790,6 +792,37 @@ template void explain_unary_operation_single_output_error<float16>(
...
+  msg << "     Input bits: " << str(FPBits<InType>(input)) << '\n';
+  msg << '\n' << "  Match decimal: " << mpfrMatchValue.str() << '\n';
+  msg << "     Match bits: " << str(FPBits<OutType>(matchValue)) << '\n';
+  msg << '\n' << "    MPFR result: " << mpfr_result.str() << '\n';
+  msg << "   MPFR rounded: " << str(FPBits<OutType>(mpfr_result.as<OutType>()))
+      << '\n';
+  msg << '\n';
+  msg << "      ULP error: " << mpfr_result.ulp_as_mpfr_number(matchValue).str()
+      << '\n';
+  tlog << msg.str();
```
Comment (overmighty) [inline libc/utils/MPFRWrapper/MPFRUtils.cpp:817]: Not sure if I should just undo this for now or fix the `msg_data` part. I think it is a pretty nice quality-of-life improvement as it means no longer having to disable multi-threading in exhaustive_test.h to get readable `explain_*_error` output, although there is still some overlap in test logs even with this (I haven't looked into it yet).

File: libc/src/__support/FPUtil/generic/sqrt.h (hunk 3 above)
Comment (lntue) [inline libc/src/__support/FPUtil/generic/sqrt.h:20]: Use proxy header.

File: libc/test/src/math/exhaustive/exhaustive_test.h [hunk 5]
```diff
@@ -68,6 +68,41 @@ struct UnaryOpChecker : public virtual LIBC_NAMESPACE::testing::Test {
...
+    do {
+      FPBits xbits(bits);
+      FloatType x = xbits.get_val();
+      bool correct =
+          TEST_MPFR_MATCH_ROUNDING_SILENTLY(Op, x, FUNC(x), 0.5, rounding);
+      failed += (!correct);
+      // Uncomment to print out failed values.
+      if (!correct) {
+        EXPECT_MPFR_MATCH_ROUNDING(Op, x, FUNC(x), 0.5, rounding);
+      }
```
Comment (overmighty) [inline libc/test/src/math/exhaustive/exhaustive_test.h:100]: I forgot to re-comment this.

File: libc/utils/MPFRWrapper/MPFRUtils.cpp [hunk 6]
```diff
@@ -790,6 +792,37 @@ template void explain_unary_operation_single_output_error<float16>(
     Operation op, float16, float16, double, RoundingMode);
 #endif
 
+template <typename OutType, typename InType>
+void explain_unary_narrower_operation_single_output_error(
```
Comment (lntue) [inline libc/utils/MPFRWrapper/MPFRUtils.cpp:796]: Maybe completely replacing `explain_unary_operation_single_output_error` with this version?

File: libc/utils/MPFRWrapper/MPFRUtils.cpp [hunk 7]
```diff
@@ -974,6 +1007,22 @@ template bool compare_unary_operation_single_output<float16>(Operation, float16,
                                                              RoundingMode);
 #endif
 
+template <typename OutType, typename InType>
+bool compare_unary_narrower_operation_single_output(Operation op, InType input,
```
Comment (lntue) [inline libc/utils/MPFRWrapper/MPFRUtils.cpp:1011]: Maybe change `compare_unary_operation_single_output` to 2 template parameters instead.

File: libc/utils/MPFRWrapper/MPFRUtils.h [hunk 8]
```diff
@@ -271,7 +290,8 @@ class MPFRMatcher : public testing::Matcher<OutputType> {
 // types.
 template <Operation op, typename InputType, typename OutputType>
 constexpr bool is_valid_operation() {
-  return (Operation::BeginUnaryOperationsSingleOutput < op &&
+  return (op == Operation::Sqrt) ||
```
Comment (lntue) [inline libc/utils/MPFRWrapper/MPFRUtils.h:293]: Add a separate condition like:
```
constexpr bool IS_NARROWING_OP = (op == Operation::Sqrt) && ....
```

File: libc/utils/MPFRWrapper/MPFRUtils.cpp [hunk 9]
```diff
@@ -757,26 +757,32 @@ ternary_operation_one_output(Operation op, InputType x, InputType y,
...
+  msg << '\n' << "  Match decimal: " << mpfrMatchValue.str() << '\n';
+  msg << "     Match bits: " << str(FPBits<OutputType>(matchValue)) << '\n';
+  msg << '\n' << "    MPFR result: " << mpfr_result.str() << '\n';
+  msg << "   MPFR rounded: "
+      << str(FPBits<OutputType>(mpfr_result.as<OutputType>())) << '\n';
+  msg << '\n';
+  msg << "      ULP error: " << mpfr_result.ulp_as_mpfr_number(matchValue).str()
+      << '\n';
+  if (msg.overflow())
+    __builtin_unreachable();
```
Comment (overmighty) [inline libc/utils/MPFRWrapper/MPFRUtils.cpp:784]: The buffer size of 1024 should always be much more than enough.

File: libc/utils/MPFRWrapper/MPFRUtils.cpp [hunk 10]
```diff
@@ -790,37 +796,10 @@ template void explain_unary_operation_single_output_error<long double>(
...
-}
-
-#ifdef LIBC_TYPES_HAS_FLOAT16
-template void
-explain_unary_narrower_operation_single_output_error<float16, float>(
-    Operation op, float, float16, double, RoundingMode);
+template void explain_unary_operation_single_output_error<float>(Operation op,
+                                                                 float, float16,
+                                                                 double,
+                                                                 RoundingMode);
```
Comment (overmighty) [inline libc/utils/MPFRWrapper/MPFRUtils.cpp:802]: The `<float>` part can be deduced, so I don't know if I should leave it as is, or only change it to `<float, float16>` here, or also change it to `<float16, float16>`, etc, elsewhere, or remove the explicit parameters everywhere and let the compiler deduce them.

File: libc/utils/MPFRWrapper/MPFRUtils.cpp [hunk 11]
```diff
@@ -757,26 +757,32 @@ ternary_operation_one_output(Operation op, InputType x, InputType y,
...
+template <typename InputType, typename OutputType>
+void explain_unary_operation_single_output_error(Operation op, InputType input,
+                                                 OutputType matchValue,
                                                  double ulp_tolerance,
                                                  RoundingMode rounding) {
-  unsigned int precision = get_precision<T>(ulp_tolerance);
+  unsigned int precision = get_precision<InputType>(ulp_tolerance);
   MPFRNumber mpfrInput(input, precision);
   MPFRNumber mpfr_result;
   mpfr_result = unary_operation(op, input, precision, rounding);
```
Comment (overmighty) [inline libc/utils/MPFRWrapper/MPFRUtils.cpp:768]: Not written by me, but the naming/case isn't consistent, and the assignment could be merged with the declaration.

File: libc/src/__support/FPUtil/generic/sqrt.h [hunk 12]
```diff
@@ -120,47 +132,119 @@ LIBC_INLINE cpp::enable_if_t<cpp::is_floating_point_v<T>, T> sqrt(T x) {
...
+          case FE_TONEAREST:
+            if (rb)
+              return OutFPBits::inf().get_val();
+            return OutFPBits::max_normal().get_val();
+          case FE_UPWARD:
+            return OutFPBits::inf().get_val();
+          default:
+            return OutFPBits::max_normal().get_val();
+          }
+        }
```
Comment (overmighty) [inline libc/src/__support/FPUtil/generic/sqrt.h:196]: This is actually not needed, I'll remove it.

File: libc/test/src/math/exhaustive/exhaustive_test.h [hunk 13]
```diff
@@ -68,6 +68,41 @@ struct UnaryOpChecker : public virtual LIBC_NAMESPACE::testing::Test {
   }
 };
 
+template <typename OutType, typename InType>
+using UnaryNarrowingOp = OutType(InType);
+
+template <typename OutType, typename InType, mpfr::Operation Op,
+          UnaryNarrowingOp<OutType, InType> Func>
+struct UnaryNarrowingOpChecker : public virtual LIBC_NAMESPACE::testing::Test {
```
Comment (lntue) [inline libc/test/src/math/exhaustive/exhaustive_test.h:76]: Can we combine this and `UnaryOpChecker` into 1?

File: libc/src/math/generic/CMakeLists.txt [hunk 14]
```diff
@@ -3601,3 +3601,15 @@ add_entrypoint_object(
...
 )
+
+add_entrypoint_object(
+  f16sqrtf
+  SRCS
+    f16sqrtf.cpp
+  HDRS
+    ../f16sqrtf.h
+  DEPENDS
+    libc.src.__support.FPUtil.sqrt
```
Comment (lntue) [inline libc/src/math/generic/CMakeLists.txt:3612]: also `libc.src.__support.macros.properties.types`

Comment (overmighty) [reply libc/src/math/generic/CMakeLists.txt:3613]: I'll open a PR later to fix it for the other `*f16` functions above too.

File: libc/utils/MPFRWrapper/MPFRUtils.cpp (hunk 10 above)
Comment (lntue) [reply libc/utils/MPFRWrapper/MPFRUtils.cpp:802]: probably you can remove `<input type>` from the all the instantiations if they can be deduced.

**Review by lntue — APPROVED** [review]

**Pull request merged**